### DISCUSSION TITLE
Built-in matching braces highlightiing

### DISCUSSION THREAD

**Discussion by pyscripter** [TurboPack/SynEdit#207] [Ideas]
I propose to add matching-braces highlighting functionality into SynEdit, implemented with indicators:

-  TSynCustomHighlighter gets two attributes MatchingBraceAttri, UnbalancedBraceAttri and a Braces property ('()[]{}<>')
- TSynEditorOption gets a new member eoHighlightMatchingBraces
- TSynEdit.SetCaretXYEx and TSynEdit.LinesChanged clear existing matching brace indicators and check for new ones if a highlighter is attached and eoHighlightMatchingBraces is set. 

Given, that we use highlighter attributes, this would allow to set Foreground, Background color and font style for matching braces, but not say painting a rectangle around the matching braces.

@MShark67 @JaFi-cz @dados  What do you think?

**Comment by MShark67** [comment]
I think it's a nice addition.  I just use a single foreground color change for my brace highlighting so that would meet my needs perfectly.

**Reply by JaFi-cz** [reply]
Its nice, but it solves part of the problem only. What about matching keywords?
I provide matching words like begin-end, if-endif, start/end of the HTML tag...
For it is necessary to add possibility enter list of pairs, check if pairs are keywords and not comments same like you need to do to check if bracket is really bracket attribute

**Reply by pyscripter** [reply]
Good point.   

For more complex stuff you can still use a plugin of some event handles as you do now. PaintTransient should still work.

Alternatively use Indicators
- Handle the OnChange event and StatusChange event (Caret movements).  The event handlers should remove the brace matching indicators and then if needed add new ones.

The latter would be more efficient.
